ctru-rs/src/services/ndsp/wave.rs

```diff
@@ -0,0 +1,138 @@
+use ctru_sys::ndspChnWaveBufClear;
+
+use super::AudioFormat;
+use crate::linear::LinearAllocator;
+
+/// Base struct to represent audio wave data. This requires audio format information.
+pub struct WaveBuffer {
+    /// Buffer data. This data must be allocated on the LINEAR memory.
+    data: Box<[u8], LinearAllocator>,
+    audio_format: AudioFormat,
+    nsamples: usize, // We don't use the slice's length here because depending on the format it may vary
+                     // adpcm_data: AdpcmData, TODO: Requires research on how this format is handled.
+}
+
+/// Informational struct holding the raw audio data and playaback info. This corresponds to [ctru_sys::ndspWaveBuf]
+pub struct WaveInfo<'b> {
+    /// Data block of the audio wave (plus its format information).
+    buffer: &'b mut WaveBuffer,
+    // Holding the data with the raw format is necessary since `libctru` will access it.
+    pub(crate) raw_data: ctru_sys::ndspWaveBuf,
+    played_on_channel: Option<u8>,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(u8)]
+pub enum WaveStatus {
+    Free = ctru_sys::NDSP_WBUF_FREE as u8,
+    Queued = ctru_sys::NDSP_WBUF_QUEUED as u8,
+    Playing = ctru_sys::NDSP_WBUF_PLAYING as u8,
+    Done = ctru_sys::NDSP_WBUF_DONE as u8,
+}
+
+impl WaveBuffer {
+    pub fn new(data: Box<[u8], LinearAllocator>, audio_format: AudioFormat) -> crate::Result<Self> {
+        let nsamples: usize = data.len() / (audio_format.sample_size() as usize);
+
+        unsafe {
+            let _r = ctru_sys::DSP_FlushDataCache(data.as_ptr().cast(), data.len() as u32);
+        }
+
+        Ok(Self {
+            data,
+            audio_format,
+            nsamples,
+        })
+    }
+
+    pub fn get_mut_data(&mut self) -> &mut [u8] {
+        &mut self.data
+    }
+
+    pub fn get_format(&self) -> AudioFormat {
+        self.audio_format
+    }
+
+    pub fn get_sample_amount(&self) -> usize {
+        self.nsamples
+    }
+}
+
+impl<'b> WaveInfo<'b> {
+    pub fn new(buffer: &'b mut WaveBuffer, looping: bool) -> Self {
+        let address = ctru_sys::tag_ndspWaveBuf__bindgen_ty_1 {
+            data_vaddr: buffer.data.as_ptr().cast(),
+        };
+
+        let raw_data = ctru_sys::ndspWaveBuf {
+            __bindgen_anon_1: address, // Buffer data virtual address
```

Comment:
Any way to make this field named better?

Reply:
I’m no expert in bindgen, but I think it’s needed since the original struct has a completely anonymous field (things that doesn’t exist in Rust, for good reasons). I wonder if we can make a special case in the build script to give it a name ourselves…

Reply:
https://github.com/rust-lang/rust-bindgen/issues/1971 looks like it currently isn't an option